components/Modals/JoinGameModal/index.tsx

```diff
@@ -0,0 +1,819 @@
+
+import {
+  BODY_TRAITS,
+  CLOTHING_TRAITS,
+  EYES_TRAITS,
+  formatFileName,
+  getImageUrl,
+  HAIR_TRAITS,
+  MOUTH_TRAITS,
+  TRAITS,
+  TraitType,
+} from './traits';
+
+type JoinGameModalProps = {
+  isOpen: boolean;
+  onClose: () => void;
+};
+
+export const JoinGameModal: React.FC<JoinGameModalProps> = ({
+  isOpen,
+  onClose,
+}) => {
+  const { game, character, reload: reloadGame } = useGame();
+  const { data: walletClient } = useWalletClient();
+  const publicClient = usePublicClient();
+  const toast = useToast();
+
+  const {
+    file: characterAvatar,
+    setFile: setCharacterAvatar,
+    onRemove,
+    onUpload,
+    isUploading,
+    isUploaded,
+  } = useUploadFile({ fileName: 'characterAvatar' });
+  const [step, setStep] = useState<number>(0);
+
+  const [characterName, setCharacterName] = useState<string>('');
+  const [characterDescription, setCharacterDescription] = useState<string>('');
+
+  const [showUpload, setShowUpload] = useState<boolean>(false);
+  const [activeTab, setActiveTab] = useState<TraitType>(TraitType.BODY);
+  const [traits, setTraits] = useState<string[]>([
+    '1_Basic_a',
+    '2_Basic_a',
+    '3_Bald_a',
+    '4_Villager1_a',
+    '5_Basic_a',
+  ]);
+  const [isMerging, setIsMerging] = useState<boolean>(false);
+
+  const [showError, setShowError] = useState<boolean>(false);
+  const [isCreating, setIsCreating] = useState<boolean>(false);
+  const [txHash, setTxHash] = useState<string | null>(null);
+  const [isSyncing, setIsSyncing] = useState<boolean>(false);
+  const [isSynced, setIsSynced] = useState<boolean>(false);
+
+  const invalidCharacterDescription = useMemo(() => {
+    return characterDescription.length > 200 && !!characterDescription;
+  }, [characterDescription]);
+
+  const hasError = useMemo(() => {
+    return (
+      !characterDescription ||
+      (showUpload && !characterAvatar) ||
+      !characterName ||
+      invalidCharacterDescription
+    );
+  }, [
+    characterDescription,
+    characterAvatar,
+    characterName,
+    invalidCharacterDescription,
+    showUpload,
+  ]);
+
+  const resetData = useCallback(() => {
+    setStep(0);
+    setShowUpload(false);
+    setActiveTab(TraitType.BODY);
+    setTraits([
+      '1_Basic_a',
+      '2_Basic_a',
+      '3_Bald_a',
+      '4_Villager1_a',
+      '5_Basic_a',
+    ]);
+
+    setCharacterName('');
+    setCharacterDescription('');
+    setCharacterAvatar(null);
+    setShowError(false);
+
+    setIsCreating(false);
+    setTxHash(null);
+    setIsSyncing(false);
+    setIsSynced(false);
+  }, [setCharacterAvatar]);
+
+  useEffect(() => {
+    if (!isOpen) {
+      resetData();
+    }
+  }, [resetData, isOpen]);
+
+  const mergeTraitImages = useCallback(async (): Promise<string> => {
+    try {
+      setIsMerging(true);
+      const [blob1, blob2, blob3, blob4, blob5] = await Promise.all([
+        fetch(getImageUrl(TRAITS[traits[0]])).then(r => r.blob()),
```

Comment:
Everything works great. Raelly awesome work! :fire:

I'm merging this. But IMO we shouldn't be uploading all these images to the server. 

Instead we could just send the image "filenames" or "keys" or "ids" and the server will already have the image urls and such and it can download it directly. Or if we have it locally then just use it and construct the final image. 

Wdyt @ECWireless ?

Reply:
Yes, good point!